### DISCUSSION TITLE
Write/Read Masks

### DISCUSSION THREAD

**Discussion by Tylertron1998** [SixLabors/ImageSharp.Drawing#67] [Q&A] [answered]
How does one achieve write/read masks using your api?

**Comment by antonfirsov** [comment]
@Tylertron1998 can you be more specific? This can mean many things for me.

**Reply by Tylertron1998** [reply]
for example, say I want to write a 100x100 image onto another 100x100 image, but I want only a 25x25 circle in the middle to be 'written', in other frameworks (namely, imagemagik) I would put down a mask on the image to be written *to* of a 25x25 circle, and then write, and then only any pixel that lies within the circle would be written over by the image that is being written *from*. The same with reading

**Comment by tocsoft** [comment, marked as answer]
This isn't directly possible in our current API, however you could adapt the techniques from one of our samples (linked below) where you can make use of the `ImageBrush` to fill a shape of your choosing with the image/effect you want to apply.

Process is roughly;
1. Create an `IPath` that represents your desired region.
2. Make sure you have 2 image instances, src(buffer you want draw) and destination(buffer you want to draw to) (clone destination if you are just apply an effect and not drawing a 2nd image).
3. Apply any effects etc to src as required.
4. Crop down src so it fits the `IPath.Bounds` of your region.
5. Fill your `IPath` using an `ImageBrush` taking your now cropped src in its constructor.
6. ...
7. Profit.

_Working sample applying a Gaussian Blur to only the inside of a Star_
https://github.com/SixLabors/Samples/blob/master/ImageSharp/ApplyEffectInsideShape/Program.cs

**Reply by Tylertron1998** [reply]
Awesome! Thanks for the answer. Will read/write masks ever be supported? Is this process less efficient then a typical read/write mask implementation?

**Reply by antonfirsov** [reply]
With an arbitrary `IPath` yes. A more efficient method would be probably to alter the alpha channel of `src` according to the shape you want to draw (making the non drawable, "background" pixels transparent), and draw it as a rectangular region on `destination`.
